### ISSUE TITLE
Validating LJ_LIKELY and LJ_UNLIKELY

### ISSUE THREAD

**Issue by lukego** [raptorjit/raptorjit#53]
I want to establish how beneficial the macros `LJ_LIKELY()` and `LJ_UNLIKELY()` are for RaptorJIT. My impression from other projects like the Linux kernel is that these are often unhelpful or even counter productive.

If they are not clearly helpful then I would like to remove them to reduce the cognitive burden of thinking about how to maintain the existing 127 uses and when to add more.

I ran the standard benchmarks in four different variants (30 runs each):
- `BOTH`: Compile with both LJ_LIKELY and LJ_UNLIKELY defined.
- `LIKELY`: Compile with only `LJ_LIKELY` (`LJ_LIKELY` becomes a nop.)
- `UNLIKELY`: Compile with only `LJ_UNLIKELY`.
- `NONE`: Compile without either.

I decided to visualize the results a little differently this time. I am using a boxplot to summarize the distribution of results instead of only showing the average. I made the figure large to accommodate this, please click to zoom:

![likely](https://cloud.githubusercontent.com/assets/13791/24110403/a5b18878-0d93-11e7-81b4-f6f93454b00d.png)

Observations:

There are a significant number of outliers. Overall the distribution of results is quite interesting and I think it has been much too simplistic to report only average and relative-standard-deviation previously. Have to think about the most appropriate ways going forward.

The boxplot really shows the non-deterministic performance of the `roulette` benchmark well. That's the benchmark that I wrote specifically to highlight how side-trace heuristics can make performance highly variable.

More:
- `binary_trees` gets -2.25% from `LJ_LIKELY()`.
- `nbody` gets +1% from `LJ_UNLIKELY()` and -1% from `LJ_UNLIKELY()` (performance is the same when both are enabled or both are disabled.)
- `life` does better with both disabled.

... and so on. I suppose this should be analyzed more thoroughly e.g. with something along the lines of Tukey's Test or a permutation test. Just don't have the time to work that out now.

Bottom line: Overall it does not seem like we are benefitting from the static branch prediction annotations in the benchmark suite. This could justify removing them. Certain instances could be examined closely in order to preserve the specific good cases while avoiding the specific bad cases. More data from real applications would be needed to validate any proposed action.

**Comment by lukego** [comment]
(I updated the plot above to be wider and have more legible colours.)

**Comment by lukego** [comment]
So! I want to move forwards now both on validating the (non)benefit of existing optimizations and also on improving our method of doing this. Here is one next step.

### Method

The boxplots make it clear that the _distribution of results_ is more interesting than a simple metric like the average result or the median result or the first result. So we should take a closer look at these distributions. I have rerun the tests for more data (500 iterations of each benchmark instead of 30, 24h of machine time) and I have made a couple of new visualizations of the results.

The first visualization is a "faceted jitter plot." Faceted means that each benchmark is split into a separate sub-plot. Jitter plot means that each individual benchmark results is plotted separately with the Y position showing the performance and the X position being random jitter (added to prevent the points from overlapping too much.) Click to zoom as always:

![likely-jitter](https://cloud.githubusercontent.com/assets/13791/24139692/62c67796-0e1d-11e7-99be-354f89a911b6.png)

The second is a [violin plot](https://en.wikipedia.org/wiki/Violin_plot) that draws a shape showing the relative frequency of results: the shape is wide around common results and narrow around rare ones.

![likely-violin](https://cloud.githubusercontent.com/assets/13791/24139773/be9749ba-0e1d-11e7-8abf-0399af367475.png)

There are some interesting patterns here that need to be explained:

- Certain benchmarks perform much more consistently than the others: especially `scimark-sor`, `pidigits-nogmp`, `mandelbrot`.
- Certain benchmarks have a bimodal distribution: especially `roulette` (known to be due to side-trace heuristics), `array3d`, `recursive-ack`, `life`.
- Certain benchmarks have occasional slow runs that don't follow a crisp pattern: especially `spectral-norm`, `recursive-fib`, `nsieve-bit`, `nbody`.

I would like to understand the reasons for all of these deviations. RaptorJIT is supporting soft-realtime applications where it is critical to _consistently_ meet performance targets and so I want to eliminate as many sources of uncertainty as possible. (The result distributions should all look like `scimark-sor` where each of the 500 runs took almost exactly the same time.)

### LJ_LIKELY

Coming back to the issue that we started from, if `LJ_LIKELY()` and `LJ_UNLIKELY()` annotations in the source code are beneficial then we should see a clear pattern of the red ink being higher i.e. the `BOTH` variant outperforming the others. There is no such pattern in the data and so I conclude that on the whole these are not helpful for RaptorJIT with our narrow Haswell + clang-4.0 target platform. These declarations are therefore eligible for removal. (However it would be lovely to cherry-pick any specific declarations that do help e.g. if we could have the "blue" performance on `array3d` and `nsieve` with no downside and the minimum of necessary declarations.)

**Comment by lukego** [comment]
(I am rerunning the earlier benchmarks with more iterations and I will apply the same visualizations. Likely I should adopt one of these visualizations for the standard report and even move the benchmarking-and-report-generation nix expression into the repo so that it is easy to run on your own machine.)

**Comment by lukego** [comment]
Couple of first things to look at when tracking down variation: hardware & test length.

Hardware is as predictable as I have ever been able to manage:
- Tests running on identical Hetzner EX41-SSD servers.
- Identical OS/kernel/etc on all servers (automatically built & deployed from Git with nix.)
- No other activities on the machines.
- Hyperthreading disabled.
- Intel TurboBoost disabled.

Test length: I am concerned that shorter tests may show more variability than long tests if the underlying reason is e.g. a system call taking longer due to a file cache miss. Ideally every test will be run for "long enough" to average out such things but I am not sure if that is true now. Here is a quick graph comparing the variability on the Y-axis (expressed as relative standard deviation - maybe not the best) with the duration in seconds of each test on the X-axis:

![rsd-vs-length](https://cloud.githubusercontent.com/assets/13791/24142265/53b30d4c-0e26-11e7-90cd-a573aa0a0ffa.png)

On the one hand there does seem to be a better of shorter tests exhibiting more variability, as seen in the downwards slope moving right. On the other hand there are definitely many tests that have similar durations but much different variability (the most extreme being `roulette` and `spectral-norm` but also many others.)

So! Could be an idea to subtract out the duration-based trend somehow, either with a statistical model or simply by tweaking the test parameters so that each one takes around 2 seconds if possible. Could also be that another metric than RSD will be more robust and helpful. One step at a time :).

**Comment by agladysh** [comment]
@lukego Side question: what do you use to generate these diagrams? :)

**Comment by lukego** [comment]
@agladysh This is with [ggplot2](http://ggplot2.org/).

Once I have a moment I will add the diagram generation to the main repo and then it will be a one-liner to run the benchmarks and generate the diagrams. (Thanks to nix that will automatically install all the various dependencies.)

**Comment by lukego** [comment]
@agladysh Here is an example: The standard benchmark report as [Rmarkdown source](https://github.com/raptorjit/branch-tests/blob/master/benchmark-results.Rmd) and then as a [rendered document](https://hydra.snabb.co/build/1100241/download/2/benchmark-results.html) created automatically by CI.

Good jumping in point for mucking around is [R Graphics Cookbook](http://www.cookbook-r.com/Graphs/).

**Comment by Laaas** [comment]
Kind of related: Maybe RaptorJIT could find out which interpreted paths are common, to do the equivalent of LJ_(UN)LIKELY automatically? You could e.g. switch which branches are the "non-default", e.g. jumped to via a Jcc instruction, and which is the fall-through case, since the fall-through case is faster on most architectures, including Skylake. Untaken branches, i.e. the throughput for taken branches is one branch every 1 or 2 clock cycles, depending on density, and for untaken branches it is always 2 per clock cycles, according to agner.

**Comment by lukego** [comment]
@Laaas Interesting idea!

Meta-issue: the Optimization Guide (#68) says not to use the interpreter at all in performance-sensitive applications, so maybe interpreter optimizations in general have too low return on investment to bother with? (Does anybody _really_ care about a 10% interpter speedup?)

I suppose that the JIT is already doing this kind of optimization i.e. we expect the code to stay on-trace and that corresponds with the untaken branch (i.e. branch is only taken when guard fails.)

**Comment by Laaas** [comment]
Yeah you're right, the JIT-compiler is already doing that.
